### DISCUSSION TITLE
Is it possible to dynamically create query descriptions based on a list of types?

### DISCUSSION THREAD

**Discussion by proc-gen** [genaray/Arch#89] [Q&A] [answered]
My implementation seems like I'm really close, but not quite there and may be something I just don't know how to do.

Essentially, I have components that all inherit from an interface and I'd like to generate query descriptions based on a list of types.

```
var type = typeof(ISpell);
var types = AppDomain.CurrentDomain.GetAssemblies()
    .SelectMany(s => s.GetTypes())
    .Where(p => type.IsAssignableFrom(p) && !p.IsInterface);

foreach (var spellType in types)
{
    var spellQuery = new QueryDescription();
    spellQuery.All.Append(Component.GetComponentType(spellType));
    spellQueries.Add(spellQuery);
    componentTypes.Add(spellType);
}

for(int i = 0; i < spellQueries.Count; i++)
{
    var spellQuery = spellQueries[i];
    world.Query(in spellQuery, (in Entity entity, ref Position pos) =>
    {
        var spell = entity.Get(componentTypes[i]);
        ((ISpell)spell).Cooldown = processSpell(gameTime, entity, pos, (ISpell)spell);
        entity.Set(spell);
    });
}
```

When I try to do this, I get the following error:
```
Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
```
My guess is that it has to do with the way I'm passing `spellQuery` into `world.Query`.

Something like this would make it very easy for me to automatically have new spell components processed correctly, for example if I added Spell4, Spell5, and Spell6 in addition to the existing Spell1, Spell2, and Spell3. Same thing would apply with components I use for effects like Burn, Shock, and Slow.

**Comment by genaray** [comment]
Is this commited already? If so i would check it out real quick ^^ 

I assume that theres a conversion error somewhere. What type is the array ComponentTypes? 

My assumption, it is : "Type[]" which is then being passed into the get? If so thats not valid and would produce such an error ^^

**Reply by proc-gen** [reply]
It made it through all the lines inside the query and it highlighted the entire `world.Query` when it threw the exception.

I just re-added it locally, but didn't check it in because of other things I'm working on. I updated it so the lists are the following:

```
List<QueryDescription> spellQueries = new List<QueryDescription>();
List<ComponentType> componentTypes = new List<ComponentType>();
```

And now populated via:
```
foreach (var spellType in types)
{
    var spellQuery = new QueryDescription();
    spellQuery.All.Append(Component.GetComponentType(spellType));
    spellQueries.Add(spellQuery);
    componentTypes.Add(Component.GetComponentType(spellType));
}
```

The reason I think it's passing `spellQuery` into `world.Query` is because it highlights the entire query for the exception and not a specific line.

![image](https://user-images.githubusercontent.com/65076703/232904211-a1e3b2ab-eb26-418f-8924-f6a41b787073.png)

**Comment by genaray** [comment]
Can you take a look at the "entity.Get()" method? Ts signature? Sometimes it happens that e.g. the generic methods are favored which is not what we want in this case ^^

The highlighting also occurs once an error inside that method is called, since most code gets inlined :)

Would be great if you could check it in on an other branch, that way i could debug it... If its an bug :)

**Reply by proc-gen** [reply]
I threw it up in this branch: https://github.com/proc-gen/roguelite-survivor/tree/generic-query-descriptions

Everything is in `~\Systems\AttackSystem.cs`, and it'll fail as soon as you start a level. Hopefully you find something I'm not seeing :)

**Reply by genaray** [reply]
Gonna do that once im home! :)

**Comment by genaray** [comment, marked as answer]
Sooo i found the issue. Its not a bug, it was an logic mistake ^^ 

First of all, your `spellQueries` are... empty. No all, any or none set. I assume that your code to fill the doesn't work properly.
Thus each SpellQuery targets ALL entities.

Second, the `world.Query` trys to access `ref Position position`. Since the query targets ALL entities, it tries to receive the position from ALL of them (also entities without it) which throws that exception.

Use this instead for your spellQueries : 
```csharp
                var spellQuery = new QueryDescription
                {
                    All = new[] { Component.GetComponentType(spellType), typeof(Position) }
                };
```

So make sure the spellQueries are filling correctly and its fine ^^

**Reply by proc-gen** [reply]
I don't use arrays very often, so I guess my issue is that I assumed that using `Append` would add the component types I wanted to the existing array. I just looked at the function definition to find that it returns a reference to the new array, which explains why the queries were empty in my original version.

Edit: Just got these changes include in the main branch along with similar updates for status effects
